### PR TITLE
Enable WASM metering and add benchmarks

### PR DESCRIPTION
**Pull request by lostman**
### Description

This PR adds an option to enable WASM metering middleware in `WasmIndexExecutor`. The metering assigns a cost of one point to each opcode. Executor is supplied with points, and when these run out, the execution is stopped.

This is intended to replace the `tokio::time::timeout`, which can't interrupt `spawn_blocking` threads, or stop WASM modules in an infinite loop.

### Testing steps

To run `criterion` benchmarks:
```
cargo bench
```
To also run Rust's benchmarks:
```
cargo bench --features benchmarks
```
The feature flag is required because benchmarks are still an unstable feature 🤷‍♂️.

I ran into issues with `criterion`, so I'd like to have both for now.

After some tweaking, the results from both benchmarks align.

#### fuel-indexer with metering

```
cargo run -p fuel-indexer -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml --indexer-handler-metering-points 100000000 --replace-indexer
```

will fail with:

```
2023-07-06T08:47:33.465076Z ERROR fuel_indexer::executor: 386: Indexer executor run time limit exceeded. Giving up. <('.')>. Consider increasing metering points
```

after encountering a large block.

Increasing `--indexer-handler-metering-points 10000000000` let's `fuel-indexer` run smoothly.

#### looping executors

Another test case is submitting an executor with a `loop {}`. For instance, changing the `fuel-indexer-test` executor to:

```rust
mod fuel_indexer_test {

    fn fuel_indexer_test_blocks(block_data: BlockData) {
        loop {}
    }
}
```

Deploying the above should result in
```
2023-07-07T14:10:19.239462Z ERROR fuel_indexer::executor: 387: Indexer executor run time limit exceeded. Giving up. <('.')>. Consider increasing metering points
```

### Changelog

* Add an option to enable metering middleware.
* Replace timeout with metering.
* Add criterion benchmarks.
* Add ordinary Rust benchmarks.

### CODE REVIEW

File: packages/fuel-indexer-benchmarks/benches/wasm_metering_benchmark.rs
Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/benches/wasm_metering_benchmark.rs:30]: Can we get an actual error message here?

File: packages/fuel-indexer-benchmarks/src/lib.rs
Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/src/lib.rs:72]: What is this magic number?

File: packages/fuel-indexer-lib/src/config/cli.rs
Comment (ra0x3) [inline packages/fuel-indexer-lib/src/config/cli.rs:171]: - I like the descriptiveness of the naming here, but let's shorten this. Maybe `max_metering_points` ? 
  - The fact that these points are for an indexer handler are implied (and mentioned explicitly in the opt description)

File: packages/fuel-indexer-lib/src/config/cli.rs
Comment (ra0x3) [inline packages/fuel-indexer-lib/src/config/cli.rs:169]: Let's also remove this magic number as well. We use defaults

File: packages/fuel-indexer-lib/src/defaults.rs
Comment (ra0x3) [inline packages/fuel-indexer-lib/src/defaults.rs:116]: Nice! Let's use this default everywhere instead of any magic numbers

File: packages/fuel-indexer-tests/tests/integration/service.rs
Comment (ra0x3) [inline packages/fuel-indexer-tests/tests/integration/service.rs:31]: 🏅

File: packages/fuel-indexer-tests/tests/integration/service.rs
Comment (ra0x3) [inline packages/fuel-indexer-tests/tests/integration/service.rs:96]: Can we get an error message here?

File: packages/fuel-indexer/src/executor.rs
Comment (ra0x3) [inline packages/fuel-indexer/src/executor.rs:679]: @lostman Going forward, we're trying to add `rustdoc` comments to any new functions we create or any previous functions/methods that we change. Can we do that here?

Makes the code much more easy to follow for others.

File: packages/fuel-indexer/src/executor.rs
Comment (ra0x3) [inline packages/fuel-indexer/src/executor.rs:743]: I think this can be changed a bit to be more readable. This `if` extends for 4 lines 😅

File: packages/fuel-indexer-benchmarks/benches/wasm_metering_benchmark.rs
Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/benches/wasm_metering_benchmark.rs:76]: More magic numbers here.

File: packages/fuel-indexer-benchmarks/benches/wasm_metering_benchmark.rs
Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/benches/wasm_metering_benchmark.rs:41]: Magic 🪄

**Review by ra0x3 — CHANGES_REQUESTED** [review]
@lostman 
- Nice work here! Testing steps work as described
- Left a few comments regarding style/form
- This code, while it works, seems a bit thrown-together -- I realize this is just for benchmarks but let's try to clean some of this stuff up (no magic numbers, add rustdoc comments, try to unwrap as little as possible, specific/descriptive error messages -- for the person who might have to work in this part of the codebase later :)

Again, works fine (one failing test on CI?). Nice work

File: packages/fuel-indexer-benchmarks/src/lib.rs
Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/src/lib.rs:63]: @lostman Why does this only go from `0..16`? More magic numbers?

File: packages/fuel-indexer-lib/src/config/cli.rs
Comment (ra0x3) [inline packages/fuel-indexer-lib/src/config/cli.rs:171]: @lostman Don't forget to update `config.yaml` with this new variable (might have to replace the old one) as well. I've created an issue to make sure we add CI to detect this automatically

File: packages/fuel-indexer-lib/src/config/mod.rs
Comment (ra0x3) [inline packages/fuel-indexer-lib/src/config/mod.rs:319]: Nice 👌🏽  but I don't see any changes to `config.yaml`. See my comment above

File: packages/fuel-indexer/src/executor.rs
Comment (ra0x3) [inline packages/fuel-indexer/src/executor.rs:500]: Unrelated to this PR, but as you previously mentioned `Manifest::graphql_schema` should produce a `GraphQLSchema` from which we can get the `version` by just calling `GraphQLSchema::version`.So this could eventually be updated to:

```rust
let version = manifest.graphql_schema().version();
```

**Review by ra0x3 — COMMENTED** [review]
@lostman 
- Left 2 pieces of feedback related to updating config.yaml and some possible more magic numbers
- @deekerno Will wait for you to review this as well - let's make sure we review the docker components as well
  - This change contains some non-trivial lower-level changes, so wanna make sure everything works
  - The change looks ok to me but better thorough than sorry

File: packages/fuel-indexer-benchmarks/src/lib.rs
Comment (deekerno) [inline packages/fuel-indexer-benchmarks/src/lib.rs:67]: Nice. You can also consider returning a closure similar to what I did in my micro-benchmarking branch: https://github.com/FuelLabs/fuel-indexer/compare/develop...deekerno/1034-add-benchmarking#diff-d85d40698746760eb4b7a4927f8f10044bc118408b876197a42d4146425d38bfR56-R93

File: packages/fuel-indexer/src/ffi.rs
Comment (deekerno) [inline packages/fuel-indexer/src/ffi.rs:241]: Is there some significance to the 1M number here? It seems like it may not be necessary for `dealloc_fn` as I can't see that costing too much; additionally, we set it immediately after the call. Also, is there a reason that we get the remaining points but don't use it until after the deallocation?

File: packages/fuel-indexer/src/executor.rs
Comment (deekerno) [inline packages/fuel-indexer/src/executor.rs:785]: Just to be clear: Reaching the metering limit results in `TrapCode::UnreachableCodeReached` and never anything else, yeah?

File: packages/fuel-indexer/src/executor.rs
Comment (ra0x3) [reply packages/fuel-indexer/src/executor.rs:576]: @deekerno Sounds like a great issue to create (as an FYI or future piece of work so we don't forget).

File: packages/fuel-indexer/src/ffi.rs
Comment (ra0x3) [reply packages/fuel-indexer/src/ffi.rs:241]: I missed this earlier, but @lostman #NoMagicNumbers - very very rarely should be including raw values in production code :) - and if you see us doing this elsewhere in the repo bring it to our attention so we can correct it

**Comment by ra0x3** [comment]
@lostman Last piece of feedback from me on this. Don't forget to update the deployment with this new WASM metering option (and remove the old option). `fuel-indexer-deployment.yaml`

**Review by deekerno — APPROVED** [review]

**Review by ra0x3 — APPROVED** [review]
LGTM. Will defer to @deekerno. Thanks for this @lostman

File: packages/fuel-indexer-benchmarks/Cargo.toml
Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/Cargo.toml:3]: @lostman If this is a non-publishable crate, this should be `0.0.0`

Comment (ra0x3) [inline packages/fuel-indexer-benchmarks/Cargo.toml:3]: @lostman If this is a non-publishable crate, this should be `0.0.0`

Here's what's in our `fuel-indexer-tests` manifest.


```toml
name = "fuel-indexer-tests"
version = "0.0.0"
edition = "2021"
license = "BUSL-1.1"
publish = false
```

**Review by ra0x3 — CHANGES_REQUESTED** [review]

**Review by ra0x3 — APPROVED** [review]

**Review by deekerno — APPROVED** [review]

**Pull request merged**
